service/mapper/transitions_test.go

```diff
@@ -0,0 +1,1233 @@
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package mapper
+
+import (
+	"io"
+	"sync"
+	"testing"
+
+	"github.com/rs/zerolog"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/optakt/flow-dps/models/dps"
+	"github.com/optakt/flow-dps/testing/mocks"
+)
+
+// For path/payload/trie/trie update test variables, see testing globals defined in pathings_test.go.
+var testLog = zerolog.New(io.Discard)
+
+func TestNewTransitions(t *testing.T) {
+	t.Run("nominal case, without options", func(t *testing.T) {
+		load := &mocks.Loader{}
+		chain := &mocks.Chain{}
+		feed := &mocks.Feeder{}
+		index := &mocks.Writer{}
+
+		tr := NewTransitions(testLog, load, chain, feed, index)
+
+		assert.NotNil(t, tr)
+		assert.Equal(t, load, tr.load)
+		assert.Equal(t, chain, tr.chain)
+		assert.Equal(t, feed, tr.feed)
+		assert.Equal(t, index, tr.index)
+		assert.NotNil(t, tr.once)
+		assert.Zero(t, tr.cfg)
+	})
+
+	t.Run("nominal case, with options", func(t *testing.T) {
+		load := &mocks.Loader{}
+		chain := &mocks.Chain{}
+		feed := &mocks.Feeder{}
+		index := &mocks.Writer{}
+
+		tr := NewTransitions(testLog, load, chain, feed, index,
+			WithIndexCommit(true),
+			WithIndexEvents(true),
+			WithIndexHeader(true),
+			WithIndexPayloads(true),
+			WithIndexTransactions(true),
+		)
+
+		assert.NotNil(t, tr)
+		assert.Equal(t, load, tr.load)
+		assert.Equal(t, chain, tr.chain)
+		assert.Equal(t, feed, tr.feed)
+		assert.Equal(t, index, tr.index)
+		assert.NotNil(t, tr.once)
+
+		assert.NotZero(t, tr.cfg)
+		assert.True(t, tr.cfg.IndexTransactions)
+		assert.True(t, tr.cfg.IndexEvents)
+		assert.True(t, tr.cfg.IndexHeader)
+		assert.True(t, tr.cfg.IndexPayloads)
+		assert.True(t, tr.cfg.IndexTransactions)
+	})
+}
+
+func TestTransitions_BootstrapState(t *testing.T) {
+	testHeight := uint64(42)
+
+	testTrie, err := trie.NewMTrie(testRoot)
+	require.NoError(t, err)
+
+	forest := &mocks.Forest{
+		SaveFunc: func(_ *trie.MTrie, _ []ledger.Path, _ flow.StateCommitment) {},
+	}
+	chain := &mocks.Chain{
+		RootFunc: func() (uint64, error) {
+			return testHeight, nil
+		},
+	}
+	load := &mocks.Loader{
+		CheckpointFunc: func() (*trie.MTrie, error) {
+			return testTrie, nil
+		},
+	}
+
+	tr := &Transitions{
+		load:  load,
+		chain: chain,
+	}
+
+	t.Run("nominal case", func(t *testing.T) {
+		t.Parallel()
+
+		// Copy state in local scope so that we can override its SaveFun without impacting other
+		// tests running in parallel.
+		var saveCalled bool
+		f := &mocks.Forest{
+			SaveFunc: func(tree *trie.MTrie, paths []ledger.Path, parent flow.StateCommitment) {
+				if !saveCalled {
+					assert.True(t, tree.IsEmpty())
+					assert.Nil(t, paths)
+					assert.Zero(t, parent)
+					saveCalled = true
+					return
+				}
+				assert.NotZero(t, tree)
```

Comment:
Isn't this just `assert.NotNil(t, tree)`? Since zero value on a pointer is `nil`? Maybe we want to `assert.False(t, tree.IsEmpty())`? Might event want to check that `assert.Len(t, tree.AllPayloads(), len(paths))`, since we need to be able to retrieve one payload per path later.